### PR TITLE
Fix clippy warnings and enable clippy in CI

### PR DESCRIPTION
**Pull request by gibbz00**
Depends on #1007.

### CODE REVIEW

File: prost-types/src/any.rs [hunk 1]
```diff
@@ -20,16 +20,13 @@ impl Any {
     {
         let expected_type_url = M::type_url();
 
-        match (
+        if let (Some(expected), Some(actual)) = (
```
Comment (caspermeijn) [inline prost-types/src/any.rs:23]: Could this be further simplified?
```suggestion
        if TypeUrl::new(&expected_type_url) == TypeUrl::new(&self.type_url) {
```

File: src/encoding.rs [hunk 2]
```diff
@@ -1681,10 +1680,11 @@ mod test {
...
     #[test]
     fn varint_overflow() {
-        let u64_max_plus_one: &[u8] = &[0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x02];
+        let mut u64_max_plus_one: &[u8] =
+            &[0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x02];
 
-        decode_varint(&mut u64_max_plus_one.clone()).expect_err("decoding u64::MAX + 1 succeeded");
-        decode_varint_slow(&mut u64_max_plus_one.clone())
+        decode_varint(&mut u64_max_plus_one).expect_err("decoding u64::MAX + 1 succeeded");
+        decode_varint_slow(&mut u64_max_plus_one)
```
Comment (caspermeijn) [inline src/encoding.rs:1687]: These `clone()` are on purpose. See https://github.com/tokio-rs/prost/pull/977

File: src/encoding.rs [hunk 3]
```diff
@@ -1616,11 +1616,10 @@ mod test {
 
             assert_eq!(encoded_len_varint(value), encoded.len());
 
-            let roundtrip_value = decode_varint(&mut encoded.clone()).expect("decoding failed");
+            let roundtrip_value = decode_varint(&mut &*encoded).expect("decoding failed");
             assert_eq!(value, roundtrip_value);
 
-            let roundtrip_value =
-                decode_varint_slow(&mut encoded.clone()).expect("slow decoding failed");
+            let roundtrip_value = decode_varint_slow(&mut &*encoded).expect("slow decoding failed");
```
Comment (caspermeijn) [inline src/encoding.rs:1622]: The re-borrow was replaced removed. See #977

**Review by caspermeijn — CHANGES_REQUESTED** [review]
I am sorry to say this, but this too many different types of changes to the code base for me to properly grasp. I find it difficult to review this PR.

Please split the changes into parts for each type of change. That could be separated commits or separated PR.

For example split all `#[cfg(not(feature = "std"))]` changes into one commit and all the `&(ref x)` changed into another commit.

**Comment by gibbz00** [comment]
> I am sorry to say this, but this too many different types of changes to the code base for me to properly grasp. I find it difficult to review this PR.

It's okay. `git add --patch` is our friend for this :) Fixed.

**Review by caspermeijn — APPROVED** [review]
Thank you for splitting the commits.

I changed the PR title to better represent the work you did.

File: src/encoding.rs [hunk 4]
```diff
@@ -1679,13 +1682,18 @@ mod test {
...
-        let u64_max_plus_one: &[u8] = &[0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0x02];
+        let mut copy = U64_MAX_PLUS_ONE;
+        decode_varint(&mut copy).expect_err("decoding u64::MAX + 1 succeeded");
+    }
 
-        decode_varint(&mut u64_max_plus_one.clone()).expect_err("decoding u64::MAX + 1 succeeded");
-        decode_varint_slow(&mut u64_max_plus_one.clone())
-            .expect_err("slow decoding u64::MAX + 1 succeeded");
+    #[test]
+    fn variant_slow_overflow() {
```
[the remark this answers is not in the document]
Comment (gibbz00) [reply src/encoding.rs:1694]: Thanks! Helps with debugability if one of them fails too ;)

**Review by caspermeijn — APPROVED** [review]

**Pull request merged**